sql: also handle ESCAPE characters properly in COPY machine

otan opened the pull request:
Release justification: critical bug fix to existing functionality

Release note (bug fix): Previously, escaping a double quote (`"`) with
`COPY` in `CSV` mode could ignore all subsequent lines in the same COPY
if an `ESCAPE` clause were specified. This is now resolved.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86929)
<!-- Reviewable:end -->

ajstorm reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86929)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[`pkg/sql/copy.go` line 538 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86929#-NAPLUUB5y2NiJwuGeiz:-NAPLUUB5y2NiJwuGej-:b28w27e) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c7921c2d7763ec76f5b372d81b92356fe47ec14/pkg/sql/copy.go#L538)):*
> ```Go
> 		} else {
> 			// Otherwise, we have to do a more involved count of quotes and
> 			// ignore any escaped quotes for counting.
> ```

Nit: Can we build out this comment a bit more, perhaps with a short example?  It took me a bit to understand what's going on below, and I worry others who encounter this code for the first time will also be confused.
___
*[`pkg/sql/copy.go` line 541 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86929#-NAPM2rMDDOzdqWIOXtg:-NAPM2rMDDOzdqWIOXth:b-309cy4) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c7921c2d7763ec76f5b372d81b92356fe47ec14/pkg/sql/copy.go#L541)):*
> ```Go
> 			isEscaped := false
> 			for _, ch := range line {
> 				if isEscaped {
> ```

Nit: I'm wondering if this would be more readable if you called this variable `skipNextChar` and then added a comment below saying that "We found a rune which matches the specified escape value.  This means we need to skip the next character, as it's been escaped"?


<!-- Sent from Reviewable.io -->

otan committed:
sql: also handle ESCAPE characters properly in COPY machine

Release justification: critical bug fix to existing functionality

Release note (bug fix): Previously, escaping a double quote (`"`) with
`COPY` in `CSV` mode could ignore all subsequent lines in the same COPY
if an `ESCAPE` clause were specified. This is now resolved.

ajstorm reviewed (approved):
This <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> now.  I'm wondering if as a follow on item in stability we should add far more robust (randomized?) testing for parsing in COPY.  I worry that this may not be the only bug lingering. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86929#-:-NAPR0sSEsHwUCofezwl:bsk56rp)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rafiss)



<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
nice find. i don't think backports are needed since v22.2 is the first release where ESCAPE is supported.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86929#-:-NAQ5xWnEXSeGDUgPZB2:b-m5xg1m)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajstorm)



<!-- Sent from Reviewable.io -->

rafiss commented:
oh, i forgot that we backported ESCAPE

daniel-crlabs commented:
This affects customers on cockroach cloud as well, see linked ticket.

otan commented:
bors r+
